com.unity.multiplayer.mlapi/Runtime/Core/RPCQueue/RPCQueueHistoryFrame.cs

```diff
@@ -0,0 +1,226 @@
+using System.Collections.Generic;
+using MLAPI.Serialization.Pooled;
+
+namespace MLAPI
+{
+    /// <summary>
+    /// QueueHistoryFrame
+    /// All queued RPCs end up in a PooledBitStream within a QueueHistoryFrame instance.
+    /// </summary>
+    public class QueueHistoryFrame
+    {
+        public enum QueueFrameType
+        {
+            Inbound,
+            Outbound,
+        }
+
+        public uint TotalSize;
+
+        public PooledBitStream QueueStream;
+        public PooledBitWriter QueueWriter;
+        public PooledBitReader QueueReader;
+
+        public List<uint> QueueItemOffsets;
+
+        private int QueueItemOffsetIndex;
+        private FrameQueueItem CurrentQueueItem;
+        private readonly QueueFrameType queueFrameType;
+
+        long CurrentStreamSizeMark;
+
+        /// <summary>
+        /// GetQueueFrameType
+        /// Returns whether this is an inbound or outbound frame
+        /// </summary>
+        /// <returns></returns>
+        public QueueFrameType GetQueueFrameType()
+        {
+            return queueFrameType;
+        }
+
+        /// <summary>
+        /// MarkCurrentStreamSize
+        /// Marks the current size of the stream  (used primarily for sanity checks)
+        /// </summary>
+        public void MarkCurrentStreamPosition()
+        {
+            if (QueueStream != null)
+            {
+                CurrentStreamSizeMark = QueueStream.Position;
+            }
+            else
+            {
+                CurrentStreamSizeMark = 0;
+            }
+        }
+
+        public long GetCurrentMarkedPosition()
+        {
+            return CurrentStreamSizeMark;
+        }
+
+
+        /// <summary>
+        /// GetCurrentQueueItem
+        /// Internal method to get the current Queue Item from the stream at its current position
+        /// </summary>
+        /// <returns>FrameQueueItem</returns>
+        private FrameQueueItem GetCurrentQueueItem()
+        {
+            //Write the packed version of the queueItem to our current queue history buffer
+            CurrentQueueItem.QueueItemType = (RPCQueueManager.QueueItemType)QueueReader.ReadUInt16();
+            CurrentQueueItem.SendFlags = (Security.SecuritySendFlags)QueueReader.ReadUInt16();
+            QueueReader.ReadSingle();
+            CurrentQueueItem.NetworkId = QueueReader.ReadUInt64();
+
+            //NSS: For now, with inbound we are punting on reading the channel and clients for these reasons:
+            //1.) For right now, channels are not pertinent as the message is already received
+            //2.) Clients are not pertinent as a received RPC is a "single target" destination
+            if (queueFrameType == QueueFrameType.Inbound)
+            {
+                QueueReader.ReadByte();
+                QueueReader.ReadInt32();
```

Comment:
what are these? o.o

Reply:
The comment explains that they are place holder values that are not pertinent to the ingress path.
However, this needs revisiting and should have a cleaner approach (i.e. if values are not used then perhaps they are not written =or= just a new type of data structure that excludes these values).
This will be added to the new Jira task.